### ISSUE TITLE
Question about video input shape 

### ISSUE THREAD

**Issue by celestialxevermore** [ArrowLuo/CLIP4Clip#75]
Dear Author, I have another Question about video input shape. 

I examine my experiment with your well-made framework : MSVD dataset these days. 

now I understood that when the class MSVD_DataLoader(Dataset) special function, __getitem__ returns 
pairs_text, paris_amsk, _pairs_segemtn, video_video_mask repectively. 

I made up in my ipynb file for recognizing the shape of these, so I found that the shape of 'video' is 
(1, 16, 1, 3, 224, 224) : b x max_frame x pair x channel x Height x Width. 

However, When I scrutinized the .py file : module_clip.py, and modeling.py, 
the forward function in CLIP4Clip class make use of video in T x 3 x H x W. 

My first question is that, what caused the tunneling from video in 6D to video in 4D. 

Also, in the forward function in VisualTransformer, 

    def forward(self, x: torch.Tensor, video_frame=-1):
        ####Conv3D AE 20220526 #####
        x_3d = x.reshape(-1, video_frame, x.shape[-3], x.shape[-2], x.shape[-1])
        if self.linear_patch == '3d':
            assert video_frame != -1
            x_3d = x.reshape(-1, video_frame, x.shape[-3], x.shape[-2], x.shape[-1])
            x_3d = x_3d.permute(0, 2, 1, 3, 4) #
            x_3d = self.conv2(x_3d)     # shape = [*, width, frame, grid, grid]
            x_3d = x_3d.permute(0, 2, 1, 3, 4)      # shape = [*, frame, width, grid, grid]
            x = x_3d.reshape(-1, x_3d.shape[-3], x_3d.shape[-2], x_3d.shape[-1]).contiguous() # shape = [*, width, grid, grid]
        else:
            x = self.conv1(x)  # shape = [*, width, grid, grid]

        x = x.reshape(x.shape[0], x.shape[1], -1)  # shape = [*, width, grid ** 2]
        x = x.permute(0, 2, 1)  # shape = [*, grid ** 2, width]
        x = torch.cat([self.class_embedding.to(x.dtype) + torch.zeros(x.shape[0], 1, x.shape[-1], dtype=x.dtype, device=x.device), x], dim=1)  # shape = [*, grid ** 2 + 1, width]
        x = x + self.positional_embedding.to(x.dtype) 
        x = self.ln_pre(x)

        ################# TAE!!! #################### 
        x = x.permute(1, 0, 2)  # NLD -> LND
        
        x = self.transformer(x, video_frame=video_frame)
        x = x.permute(1, 0, 2)  # LND -> NLD

        # Move the three lines below to `encode_image` for entire hidden sequence
        # x = self.ln_post(x[:, 0, :])
        # if self.proj is not None:
        #     x = x @ self.proj

        return x

 In looking the code x_3d = x.reshape(-1, video_frame, x.shape[-3], x.shape[-2], x.shape[-1]),
I want to know why the shape of x_3d is 5D, not 6D. the question is similar to the first. 

Third, I wonder the reason that why the frame is composed of bs * ts. Can I get to know what the 'bs' and 'ts' means?
(It can be found also in Class CLIP4Clip, modeling.py) 

    def forward(self, input_ids, token_type_ids, attention_mask, video, video_mask=None):
        input_ids = input_ids.view(-1, input_ids.shape[-1])
        token_type_ids = token_type_ids.view(-1, token_type_ids.shape[-1])
        attention_mask = attention_mask.view(-1, attention_mask.shape[-1])
        video_mask = video_mask.view(-1, video_mask.shape[-1])

        # T x 3 x H x W
        video = torch.as_tensor(video).float()
        b, pair, bs, ts, channel, h, w = video.shape
        video = video.view(b * pair * bs * ts, channel, h, w)
        video_frame = bs * ts

        sequence_output, visual_output = self.get_sequence_visual_output(input_ids, token_type_ids, attention_mask,
                                                                         video, video_mask, shaped=True, video_frame=video_frame)

        if self.training:
            loss = 0.
            sim_matrix, *_tmp = self.get_similarity_logits(sequence_output, visual_output, attention_mask, video_mask,
                                                    shaped=True, loose_type=self.loose_type)
            sim_loss1 = self.loss_fct(sim_matrix)
            sim_loss2 = self.loss_fct(sim_matrix.T)
            sim_loss = (sim_loss1 + sim_loss2) / 2
            loss += sim_loss

            return loss
        else:
            return None

Last, In the forward function in class CLIP4Clip, I really fell thankful if you let me know what is the exact  meaning of input_ids, and token_type_ids. 

Sincerely,

**Comment by ArrowLuo** [comment]
Hi @celestialxevermore, thanks for your attention. 

The confusion may be caused by the history tests and we ignored it when we clean the released code. 
1. The `(1, 16, 1, 3, 224, 224)`  in the the dataloader is  Pair x L x T x 3 x H x W. The Pair and T will always be 1, so plz ignore them. Equally, the `pair`, `ts` will be 1 in `b, pair, bs, ts, channel, h, w` ([Line 257](https://github.com/ArrowLuo/CLIP4Clip/blob/master/modules/modeling.py#L257)). The size of the video can be seen in [Line 258](https://github.com/ArrowLuo/CLIP4Clip/blob/master/modules/modeling.py#L258)
2. The `input_ids`  is the same as in CLIP, which is the ids of the text. The `token_type_ids` is the token type, having the same meaning as in BERT, which is actually redundancy here, and plz ignore. 

Best~

**Comment by celestialxevermore** [comment]
Thank you! You are my lifesaver :)
